Jellyfin.Plugin.OpenSubtitles/OpenSubtitleDownloader.cs

```diff
@@ -99,232 +96,259 @@ public async Task<IEnumerable<RemoteSubtitleInfo>> Search(SubtitleSearchRequest
+                    hash = Util.ComputeHash(fileStream);
                 }
-                : new List<SubtitleSearchParameters>
-                {
-                    new SubtitleSearchParameters(subLanguageId, imdbid: searchImdbId),
-                    new SubtitleSearchParameters(subLanguageId, query: request.Name, imdbid: searchImdbId)
-                };
-            var parms = new List<SubtitleSearchParameters>
+            }
+            catch (IOException e)
+            {
+                throw new IOException(string.Format(CultureInfo.InvariantCulture, "IOException while computing hash for {0}", request.MediaPath), e);
+            }
+
+            var options = new Dictionary<string, string>
             {
-                new SubtitleSearchParameters(
-                    subLanguageId,
-                    movieHash: hash,
-                    movieByteSize: movieByteSize,
-                    imdbid: searchImdbId),
+                { "languages", request.TwoLetterISOLanguageName },
+                { "moviehash", hash },
+                { "type", request.ContentType == VideoContentType.Episode ? "episode" : "movie" }
             };
-            parms.AddRange(subtitleSearchParameters);
 
-            if (_rateLimitLeft == 0)
+            // If we have the IMDb ID we use that, otherwise query with the details
+            if (imdbId != 0)
             {
-                await Task.Delay(1000, cancellationToken)
-                    .ConfigureAwait(false);
+                options.Add("imdb_id", imdbId.ToString(_usCulture));
             }
-
-            var searchResponse = await OpenSubtitlesHandler.OpenSubtitles.SearchSubtitlesAsync(parms.ToArray(), cancellationToken).ConfigureAwait(false);
-
-            var searchResult = searchResponse.Item1;
-            _rateLimitLeft = searchResponse.Item2 ?? _rateLimitLeft;
-
-            if (searchResponse.Item2 != null)
+            else
             {
-                if (_rateLimitLeft <= 4)
+                if (request.ContentType == VideoContentType.Episode)
+                {
+                    options.Add("query", request.SeriesName.Length <= 2 ? $"{request.SeriesName} {request.ProductionYear}" : request.SeriesName);
+                    options.Add("season_number", request.ParentIndexNumber?.ToString(_usCulture) ?? string.Empty);
+                    options.Add("episode_number", request.IndexNumber?.ToString(_usCulture) ?? string.Empty);
+                }
+                else
                 {
-                    await Task.Delay(250, cancellationToken)
-                        .ConfigureAwait(false);
+                    options.Add("query", request.Name.Length <= 2 ? $"{request.Name} {request.ProductionYear}" : request.Name);
                 }
             }
 
-            if (searchResult is not MethodResponseSubtitleSearch subtitleSearchResult)
+            if (request.IsPerfectMatch)
```

Comment:
In my opinion yes because the subtitle editor (and automatic downloader) will still point out and prioritize best matching subtitles
![](https://cdn.mbr.pw/46766261568405515.png)

Reply:
But hashing the file is extremely slow

Reply:
The method just reads the first and last 64 KiB of the file. I benchmarked it and it took about 25 ms to compute hash of 64 movies (I ran it about 1000 times, that was the average), 0.4 ms per file doesn't sound *that* slow to me :thinking:
There's really not much I can do without opening another PR but if you believe changes are needed let me know I guess

Reply:
I didn't know that. I just remembered some reports about the plug-in being slow at one point. I do wonder how it performs over a slow network mount though